src/rest_vol_dataset.c

```diff
@@ -5013,8 +5078,26 @@ RV_json_values_to_binary_callback(char *HTTP_response, const void *callback_data
     if (ret_value >= 0 && value_buffer)
         *out_buf = value_buffer;
 
-    if (ret_value < 0 && value_buffer)
+    if (ret_value < 0 && value_buffer) {
+        H5T_class_t dtype_class = H5T_NO_CLASS;
+
+        if ((dtype_class = H5Tget_class(dtype_id)) < 0)
+            FUNC_DONE_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get datatype class");
+
+        if (dtype_class == H5T_VLEN) {
+            /* Reclaim each allocated vlen buffer */
+            hvl_t *curr = (hvl_t *)value_buffer;
+
+            while (curr->p) {
+                RV_free(curr->p);
```

Comment:
Another case here where `H5Treclaim` is probably simpler and more correct than trying to do this by hand.